mtags/src/main/scala/scala/meta/internal/mtags/GlobalSymbolIndex.scala

```diff
@@ -0,0 +1,84 @@
+package scala.meta.internal.mtags
+
+import scala.meta.io.AbsolutePath
+
+/**
+ * An index to lookup the definition of global symbols.
+ *
+ * Only indexes plain Scala and Java source files, no compilation needed.
+ */
+trait GlobalSymbolIndex {
+
+  /**
+   * Lookup the definition of a global symbol.
+   *
+   * Returns the path of the file that defines the symbol but does not include
+   * the exact position of the definition. Computing the range position of the
+   * definition is left for the client and can be done using the mtags module.
+   *
+   * @param symbol a global SemanticDB symbol. For comprehensive documentation
+   *               of how a symbol is formatted consule the specification:
+   *               https://scalameta.org/docs/semanticdb/specification.html#symbol
+   *
+   *  Examples: {{{
+   *    "scala/Option#"                 // Option class
+   *    "scala/Option."                 // Option companion object
+   *    "scala/Option#.get()."          // `get` method for Option
+   *    "scala/Predef.String#"          // String type alias in Predef
+   *    "java/lang/String#format(+1)."  // Static `format` method for strings
+   *    "java/util/Map#Entry#"          // Static inner interface.
+   *  }}}
+   * @return the definition of the symbol, if any.
+   */
+  def definition(symbol: String): Option[SymbolDefinition]
+
+  /**
+   * Add an individual Java or Scala source file to the index.
+   *
+   * @param file the absolute path to the source file, can be a path
+   *            on disk or inside of a jar/zip file.
+   * @param sourceDirectory the enclosing project source directory if
+   *                        file is on disk, used to relativize `file`.
+   *                        Can be None if file is inside a zip file
+   *                        assuming the file path is already relativized
+   *                        by that point.
+   */
+  def addSourceFile(
+      file: AbsolutePath,
+      sourceDirectory: Option[AbsolutePath]
+  ): Unit
```

Comment:
This signature means the method swallows errors. Could it be useful to report them back to the caller in case something goes wrong (e.g., the file doesn't exist, or it's not a Scala/Java file?). The caller can decide to discard it anyway.

Reply:
Currently this throws an exception for errors like tokenization failures. I'm not sure yet what the best error handling method so I'd prefer to refine it down the road once we have a clearer picture of how it will be used in the language server. Maybe the underlying implementation can accept a `Reporter` 🤔

Reply:
Ok, let's wait and see. I was thinking of something like "Either" or "Try" here, nothing fancy but still better than Unit.